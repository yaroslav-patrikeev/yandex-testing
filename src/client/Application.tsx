import React, { FC, useCallback, useState } from 'react';
import { Link, NavLink, Route, Routes } from 'react-router-dom';
import { cn } from '@bem-react/classname';

import { Home } from './pages/Home';
import { Catalog } from './pages/Catalog';
import { Product } from './pages/Product';
import { Delivery } from './pages/Delivery';
import { Contacts } from './pages/Contacts';
import { Helmet } from 'react-helmet';
import { Cart } from './pages/Cart';
import { useSelector } from 'react-redux';
import { ApplicationState } from './store';

const bem = cn('Application');

export const Application: FC = () => {
    const [collapsed, setCollapsed] = useState(true);
    const cart = useSelector((s: ApplicationState) => s.cart);

    const toggle = useCallback(() => setCollapsed(!collapsed), [setCollapsed, collapsed]);
    const hide = useCallback(() => {
        if (process.env.BUG_ID === '4') {
            setCollapsed(false);
        } else {
            setCollapsed(true);
        }
    }, [setCollapsed]);

    const count = Object.keys(cart).length;
    const cartLabel = count ? `Cart (${count})` : 'Cart';
    const navbarClass = collapsed ? 'collapse navbar-collapse' : 'navbar-collapse';

    return (
        <div className={bem()}>
            <Helmet titleTemplate='%s â€” Kogtetochka store' />
            <nav className='navbar navbar-expand-sm navbar-light bg-light'>
                <div className='container'>
                    <Link className={bem('Brand', ['navbar-brand'])} to='/'>
                        Kogtetochka store
                    </Link>
                    <button className={bem('Toggler', ['navbar-toggler'])} aria-label='Toggle navigation' onClick={toggle}>
                        <span className='navbar-toggler-icon'></span>
                    </button>
                    <div className={bem('Menu', [navbarClass])}>
                        <div className='navbar-nav'>
                            <NavLink className={({ isActive }) => (isActive ? 'active nav-link' : 'nav-link')} to='/catalog' onClick={hide}>
                                Catalog
                            </NavLink>
                            <NavLink className={({ isActive }) => (isActive ? 'active nav-link' : 'nav-link')} to='/delivery' onClick={hide}>
                                Delivery
                            </NavLink>
                            <NavLink className={({ isActive }) => (isActive ? 'active nav-link' : 'nav-link')} to='/contacts' onClick={hide}>
                                Contacts
                            </NavLink>
                            <NavLink className={({ isActive }) => (isActive ? 'active nav-link' : 'nav-link')} to='/cart' onClick={hide}>
                                {cartLabel}
                            </NavLink>
                        </div>
                    </div>
                </div>
            </nav>
            <div className='container pt-4'>
                <Routes>
                    <Route path='/' element={<Home />} />
                    <Route path='/catalog' element={<Catalog />} />
                    <Route path='/catalog/:id' element={<Product />} />
                    <Route path='/delivery' element={<Delivery />} />
                    <Route path='/contacts' element={<Contacts />} />
                    <Route path='/cart' element={<Cart />} />
                </Routes>
            </div>
        </div>
    );
};
