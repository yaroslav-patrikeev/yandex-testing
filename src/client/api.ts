import axios from 'axios';
import { CartState, CheckoutFormData, CheckoutResponse, Product, ProductShortInfo } from '../common/types';

const currentBugId = Number(process.env.BUG_ID) || undefined;

export class ExampleApi {
    constructor(private readonly basename: string) {

    }

    async getProducts() {
        return await axios.get<ProductShortInfo[]>(`${this.basename}/api/products`, { params: { bug_id: currentBugId } });
    }

    async getProductById(id: number) {
        return await axios.get<Product>(`${this.basename}/api/products/${id}`, { params: { bug_id: currentBugId } });
    }

    async checkout(form: CheckoutFormData, cart: CartState) {
        return await axios.post<CheckoutResponse>(`${this.basename}/api/checkout`, { form, cart }, { params: { bug_id: currentBugId } });
    }
}

export const LOCAL_STORAGE_CART_KEY = 'example-store-cart';

export class CartApi {
    getState(): CartState {
        try {
            const json = localStorage.getItem(LOCAL_STORAGE_CART_KEY);
            return JSON.parse(json) as CartState || {};
        } catch {
            return {};
        }
    }

    setState(cart: CartState) {
        localStorage.setItem(LOCAL_STORAGE_CART_KEY, JSON.stringify(cart));
    }
}
