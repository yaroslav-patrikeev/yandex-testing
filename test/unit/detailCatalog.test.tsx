import '@testing-library/jest-dom';

import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { InternalAxiosRequestConfig } from 'axios';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { CartApi, ExampleApi } from '../../src/client/api';
import { Application } from '../../src/client/Application';
import { initStore } from '../../src/client/store';

const basename = '/hw/store';
const api = new ExampleApi(basename);
const cart = new CartApi();

beforeEach(() => {
	jest.spyOn(api, 'getProductById').mockImplementation(() => {
		return new Promise(res => {
			res({
				data: {
					id: 0,
					name: 'Unbranded kogtetochka',
					description: 'Really Recycled kogtetochka for Chartreux',
					price: 55,
					color: 'green',
					material: 'Granite',
				},
				status: 200,
				statusText: 'OK',
				headers: {},
				config: {
					headers: {},
				} as InternalAxiosRequestConfig,
			});
		});
	});
	jest.spyOn(cart, 'getState').mockImplementation(() => {
		return {
			'0': { name: 'Rustic kogtetochka', count: 1, price: 731 },
		};
	});
});

describe('На странице c подробной информацией', () => {
	it('отображается сообщение, если товар добавлен в корзину', async () => {
		const app = (
			<MemoryRouter
				initialEntries={[`${basename}/catalog/0`]}
				basename={basename}
			>
				<Provider store={initStore(api, cart)}>
					<Application />
				</Provider>
			</MemoryRouter>
		);

		render(app);

		await waitFor(() => {
			const inCart = screen.getByText('Item in cart');
			expect(inCart).toBeInTheDocument();
		});
	});
	it('при каждом нажатии на кнопку данные добавляются в localStorage', async () => {
		const setState = jest.spyOn(cart, 'setState');
		const app = (
			<MemoryRouter
				initialEntries={[`${basename}/catalog/0`]}
				basename={basename}
			>
				<Provider store={initStore(api, cart)}>
					<Application />
				</Provider>
			</MemoryRouter>
		);

		render(app);
		await waitFor(() => {
			expect(screen.getByText('Add to Cart')).toBeInTheDocument();
		});
		const button = screen.getByText('Add to Cart');
		fireEvent.click(button);
		fireEvent.click(button);
		fireEvent.click(button);
		expect(setState).toHaveBeenCalledTimes(3);
	});
	it('верстка не изменилась', async () => {
		const app = (
			<MemoryRouter
				initialEntries={[`${basename}/catalog/0`]}
				basename={basename}
			>
				<Provider store={initStore(api, cart)}>
					<Application />
				</Provider>
			</MemoryRouter>
		);
		const { asFragment } = render(app);
		await waitFor(() => {
			expect(screen.getByText('Add to Cart')).toBeInTheDocument();
		});

		expect(asFragment()).toMatchSnapshot();
	});
});

afterEach(() => {
	jest.restoreAllMocks();
});
