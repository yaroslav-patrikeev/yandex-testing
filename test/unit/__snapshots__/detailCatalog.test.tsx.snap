// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`На странице c подробной информацией верстка не изменилась 1`] = `
<DocumentFragment>
  <div
    class="Application"
  >
    <nav
      class="navbar navbar-expand-sm navbar-light bg-light"
    >
      <div
        class="container"
      >
        <a
          class="Application-Brand navbar-brand"
          href="/hw/store"
        >
          Kogtetochka store
        </a>
        <button
          aria-label="Toggle navigation"
          class="Application-Toggler navbar-toggler"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="Application-Menu collapse navbar-collapse"
        >
          <div
            class="navbar-nav"
          >
            <a
              aria-current="page"
              class="active nav-link"
              href="/hw/store/catalog"
            >
              Catalog
            </a>
            <a
              class="nav-link"
              href="/hw/store/delivery"
            >
              Delivery
            </a>
            <a
              class="nav-link"
              href="/hw/store/contacts"
            >
              Contacts
            </a>
            <a
              class="nav-link"
              href="/hw/store/cart"
            >
              Cart (1)
            </a>
          </div>
        </div>
      </div>
    </nav>
    <div
      class="container pt-4"
    >
      <div
        class="Product"
      >
        <div
          class="ProductDetails row"
        >
          <div
            class="col-12 col-sm-5 col-lg-4"
          >
            <img
              class="Image"
              src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
            />
          </div>
          <div
            class="col-12 col-sm-7 col-lg-6"
          >
            <h1
              class="ProductDetails-Name"
            >
              Unbranded kogtetochka
            </h1>
            <p
              class="ProductDetails-Description"
            >
              Really Recycled kogtetochka for Chartreux
            </p>
            <p
              class="ProductDetails-Price fs-3"
            >
              $55
            </p>
            <p>
              <button
                class="ProductDetails-AddToCart btn btn-primary btn-lg"
              >
                Add to Cart
              </button>
              <span
                class="CartBadge text-success mx-3"
              >
                Item in cart
              </span>
            </p>
            <dl>
              <dt>
                Color
              </dt>
              <dd
                class="ProductDetails-Color text-capitalize"
              >
                green
              </dd>
              <dt>
                Material
              </dt>
              <dd
                class="ProductDetails-Material text-capitalize"
              >
                Granite
              </dd>
            </dl>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
